### DISCUSSION TITLE
Connecting to Firefly's database from Metabase or Power BI

### DISCUSSION THREAD

**Discussion by dgossghub46** [firefly-iii/firefly-iii#8620] [Firefly III] [answered]
Hello all, Firefly III looks like great product and I'm really excited to start using it.

I would like to have the flexibility to create my own visuals of the transaction data, and I'm trying to set up Metabase or PowerBI with it. I'm getting errors when trying to connect though:

Metabase: 
(for hostname = db, port = 3306): Could not connect to address=(host=db)(port=3306)(type=master) : Socket fail to connect to host:db, port:3306. db
(for hostname = <ip address>, port = 3306):  Hmm, we couldn't connect to the database. Make sure your Host and Port settings are correct

Power BI Desktop: "Microsoft SQL: A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: Named Pipes Provider, error: 40 - Could not open a connection to SQL Server)"

My set-up: HP computer running proxmox, with Firefly III in a LXC container with docker (generated using https://tteck.github.io/Proxmox/). Metabase is in a different LXC container. Power BI Desktop is on a windows machine on my local network.

Looking at the .env and .db.env files, it looks too simple: the port, username, database name, type of database (mysql) are all identified. Maybe the hostname is the problem? I used the same IP that I use to log in to firefly, which works for the firefly data importer. However the .env file provides a hostname of "db". Using "localhost" as recommended by firefly documentation never worked for me, I always had to use the IP address.

When I run "netstat -tuln" I see the ports for firefly and the data importer, but not the default port 3306. Strange, maybe PowerBI is right and it's not open; the data importer is getting in but I had to generate an OAuth client to do that. Do I need to do something special, using the API or something else?

For what it's worth the data importer runs fine.

Would appreciate any guidance on how to address this issue. Thanks!

**Comment by JC5** [comment]
You need to open a port in the database settings: `3306:3306` should do it. Then, PowerBI would have to connect to `your-ip:3306`.

Perhaps later, you have to change some settings because by default the database does not accept connections from outside the local network for security reasons. But first see if this works.

**Reply by dgossghub46** [reply]
Thanks for your quick response! I've been trying unsuccessfully to find the .conf file in which to make this change (I assume that's where I would do it?). I found the path to the database in [this post](https://docs.firefly-iii.org/how-to/firefly-iii/advanced/backup/), but I have it installed in docker and it's not obvious to me how to find the database settings to make the change you suggested. Could you please give me another nudge? Thanks,

Edit: I tried adding a "port" line to the yml, but that did not work (I rebooted the LXC container in which docker is running after saving the change to the yml). Also noticed the yml indicated paths for the app and the db, but I was unable to navigate to them, the /var/lib/mysql and /var/www did not seem to exist.

**Reply by JC5** [reply, marked as answer]
Since you're using Docker, this change would have to take place in either your `docker-compose.yml` file or your container start command. There are examples of port mappings already in the file, you could copy those.

The paths you refer to exist only inside the containers. Are you familiar with the concept?

**Reply by dgossghub46** [reply]
Thanks, works! I had forgotten to run "docker compose -f docker-compose.yml up -d --pull=always" after adding the port line to the yml.

Now I'm happy I don't have to answer your second question, and admit I'm not familiar with containers much at all :)

Thanks for your support and the great product!

**Reply by JC5** [reply]
Hah, not a problem. Glad to hear it works. Be sure to read this page about the database structure. 

https://docs.firefly-iii.org/explanation/more-information/architecture/
